Fix adj etan

timothyas opened the pull request:
## What changes does this PR introduce?
This introduces a new dummy routine which is carefully placed to correctly output of adEtaN, the adjoint variable for etaN. 

## What is the current behaviour? 
Currently, adEtaN is printed from addummy_in_stepping, but is zero at this point in the adjoint code.

## What is the new behaviour 
adEtaN is dumped out.

## Does this PR introduce a breaking change? 
no

## Other information:
Output at initial time matches the adxx_etan output from the control package.
I tested this for the various configurations of the free surface that made sense to me: nonlinear and linear free surface, `exactConserv`=t/f,  various settings for `implicDiv2DFlow` & `implicSurfPress`  ... etc. and got consistent results between the control and dumped output no matter the configuration. 

The caveat here is that I could not get this to work with the diagnostics package, and this issue is a bit technical. I put the dummy routine in this particular location so that adEtaN gets the impact of adjoint variables from `calc_r_star_ad` or `calc_surf_dr_ad`, and before it is integrated through the adjoint model, where adEtaN is zero'd out as sensitivities are passed to other variables. However, at this particular spot adEtaN is a "half" time step off from the other variables, so it did not make sense to fill the diagnostic slots for `ADJetan` to be printed later via `addummy_in_stepping` because they would be inconsistent. I also could not think of a way to print individual diagnostic variables at different times. This could be fixed by placing `dummy_in_stepping` in a similar manner to `do_state_vars_diags` but for adjoint variables, but this seemed like overkill for a first step.

christophernhill commented:
@jm-c to send post some comments on a possible way to shorten....

jm-c commented:
@timothyas this is a good job ! Just one problem and may be some simplifications:

1) There is a problem with dumpAdByRec=T case ; I think it's easy to reproduce.
 Since dumpAdRecMn is incremented twice (in addummy_in_stepping.F, the old place,
  and now a second time in dummy_for_etan.F), the reccord number in meta file 
 is funny (and *.data output file has hole in it).

Note: integr_continuity.F is called twice, in initialise_varia.F (nIter0) and 
 forward_step.F (myIter > nIter0); but I am not sure how it works in reverse mode,
 seems that integr_continuity_ad is also called from forward_step_ad with nIter0.
 Would be nice if you could check.

2) I suggest to remove some commented lines in your new S/R addummy_for_etan.F 
 (lines 136-148) and g_dummy_for_etan.F (lines 78-82), since there is no need to 
 copy/duplicate "very nasty" pieces of code.
 And if you feel like you could also remove the corresponding lines in 
 addummy_in_stepping.F (since you already cleaned-up this file from adEtaN output).

3) I think there is a way to simplify addummy_for_etan.F by just keeping the code
 corresponding to dumpAdVarExch = 0. 
In fact, in most cases, user will run with exactConserv=T and implicDiv2Dflow > 0
 so that, in model/src/integr_continuity.F (lines 306-308) EXCH is going to be applied 
 to etaN (line 308);
I would suggest, in integr_continuity.F, to ALWAYS do this EXCH call anytime 
 ALLOW_AUTODIFF_MONITOR is defined and leave the conditional EXCH call if undef, using
#else /* ALLOW_AUTODIFF_MONITOR */

With this modification (unconditional EXCH call on etaN when ALLOW_AUTODIFF_MONITOR
 is defined) we don't need to worry about AD-EXCH on adEtaN in addummy_for_etan.F
  
On Wed, Sep 12, 2018 at 04:07:47PM +0000, Timothy Smith wrote:
> ## What changes does this PR introduce?
> This introduces a new dummy routine which is carefully placed to correctly output of adEtaN, the adjoint variable for etaN. 
> 
> ## What is the current behaviour? 
> Currently, adEtaN is printed from addummy_in_stepping, but is zero at this point in the adjoint code.
> 
> ## What is the new behaviour 
> adEtaN is dumped out.
> 
> ## Does this PR introduce a breaking change? 
> no
> 
> ## Other information:
> Output at initial time matches the adxx_etan output from the control package.
> I tested this for the various configurations of the free surface that made sense to me: nonlinear and linear free surface, `exactConserv`=t/f,  various settings for `implicDiv2DFlow` & `implicSurfPress`  ... etc. and got consistent results between the control and dumped output no matter the configuration. 
> 
> The caveat here is that I could not get this to work with the diagnostics package, and this issue is a bit technical. I put the dummy routine in this particular location so that adEtaN gets the impact of adjoint variables from `calc_r_star_ad` or `calc_surf_dr_ad`, and before it is integrated through the adjoint model, where adEtaN is zero'd out as sensitivities are passed to other variables. However, at this particular spot adEtaN is a "half" time step off from the other variables, so it did not make sense to fill the diagnostic slots for `ADJetan` to be printed later via `addummy_in_stepping` because they would be inconsistent. I also could not think of a way to print individual diagnostic variables at different times. This could be fixed by placing `dummy_in_stepping` in a similar manner to `do_state_vars_diags` but for adjoint variables, but this seemed like overkill for a first step.  
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/MITgcm/MITgcm/pull/151
> 
> -- Commit Summary --
> 
>   * Remove broken calls to adEtaN
>   * New dummy routines for printing ADJetan. Added to calling sequence in the_model_main.
> 
> -- File Changes --
> 
>     M model/src/integr_continuity.F (8)
>     M model/src/the_model_main.F (1)
>     A pkg/autodiff/addummy_for_etan.F (232)
>     M pkg/autodiff/addummy_in_stepping.F (34)
>     A pkg/autodiff/dummy_for_etan.F (34)
>     A pkg/autodiff/dummy_for_etan.flow (11)
>     A pkg/autodiff/g_dummy_for_etan.F (120)
> 
> -- Patch Links --
> 
> https://github.com/MITgcm/MITgcm/pull/151.patch
> https://github.com/MITgcm/MITgcm/pull/151.diff
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/MITgcm/MITgcm/pull/151

timothyas commented:
Thanks for the comments @jm-c! 

1) I fixed this by only incrementing `dumpAdRecMn` in `addummy_for_etan.F`, since this routine is called first. Now the record numbers in *.meta are correct. Note that addetan has one more record than the other state variables simply because of where it is called: as I mentioned it's a "half" step off from the other variables. As I mentioned, only a much more intricate handling of adjoint output would fix this completely, which now that I've gone this far could be tackled (some day) if there is enough interest in the community. 
Yes, `integr_continuity.F` is called twice in reverse as well, and in the adjoint versions of the forward routines you mentioned. That's part of the reason printing from this routine works well - printing etan_ad from `initialise_varia_ad` allows the etan_ad to be printed precisely at t=0. I could not find a way to do this with only print statements in `forward_step`. 

2) Totally agreed, I removed these comments from all offending scripts :) 

3) I can't say I totally agree with this change, just because it assumes something about the user's setup, even though it should cover an overwhelming majority of cases. However, I'm not as familiar with the reasons to do (or not to do) the exchanges in `integr_continuity`, so I trust your judgement.

Lastly, it looks like my doc build failed with 299b764  but is now passing with the latest commit, 7c5219d, and locally on my machine I have no problems with `make html` or `make latexpdf`. The error is a stalled build, should I/we be concerned?

timothyas commented:
I see, I had been testing only with  adjDumpFreq = DeltaT so I didn't catch this. Now I've made a separate variable `dumpAdRecEt` for printing `adEtaN` and returned `dumpAdRecMn` back to normal. My test with the same time steps & print frequency ran just fine so hopefully that should do it. 

Also I added a line to doc/tag-index in case you guys want that (forgot originally). If it's not necessary I can revert back.

jm-c commented:
@timothyas : I simplify a bit addummy_for_etan.F, and everything looks good. 
LGTM+1

timothyas commented:
Great, thanks @jm-c!
